oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/storage/PersistenceTimer.java

```diff
@@ -92,73 +105,165 @@ public void start(ModuleManager moduleManager, CoreModuleConfig moduleConfig) {
+            // CountDownLatch makes sure all prepare threads done eventually.
+            CountDownLatch prepareStageCountDownLatch = new CountDownLatch(persistenceWorkers.size());
 
-                persistenceWorkers.forEach(worker -> {
-                    if (log.isDebugEnabled()) {
-                        log.debug("extract {} worker data and save", worker.getClass().getName());
-                    }
+            /*
+                Here we use `this.prepareRequests` as a FIFO queue, for a producer-consumer model.
+                The prepareExecutorService is for making the executable requests ready, and batchExecutorService consumes from the queue to flush.
+                When the number of metrics produced reaches maxSyncoperationNum or the prepare stage is done,
+                the data would flush into the storage.
 
-                    worker.buildBatchRequests(prepareRequests);
+                When the consumer ends or an exception occurs in the middle, the entire process is completed.
+             */
 
-                    worker.endOfRound(System.currentTimeMillis() - lastTime);
-                });
+            persistenceWorkers.forEach(worker -> {
+                prepareExecutorService.submit(() -> {
+                    if (stop.get()) {
+                        prepareStageCountDownLatch.countDown();
+                        return;
+                    }
 
-                if (debug) {
-                    log.info("build batch persistence duration: {} ms", System.currentTimeMillis() - startTime);
-                }
-            } finally {
-                timer.finish();
-            }
+                    HistogramMetrics.Timer timer = prepareLatency.createTimer();
+                    try {
+                        if (log.isDebugEnabled()) {
+                            log.debug("extract {} worker data and save", worker.getClass().getName());
+                        }
+                        List<PrepareRequest> innerPrepareRequests = new ArrayList<>(5000);
+                        worker.buildBatchRequests(innerPrepareRequests);
+                        prepareQueue.putMany(innerPrepareRequests);
+                        worker.endOfRound(System.currentTimeMillis() - lastTime);
+                    } finally {
+                        timer.finish();
+                        prepareStageCountDownLatch.countDown();
+                    }
+                });
+            });
 
-            HistogramMetrics.Timer executeLatencyTimer = executeLatency.createTimer();
-            try {
-                List<List<PrepareRequest>> partitions = Lists.partition(prepareRequests, maxSyncoperationNum);
-                CountDownLatch countDownLatch = new CountDownLatch(partitions.size());
-                for (final List<PrepareRequest> partition : partitions) {
-                    executorService.submit(() -> {
+            Future<?> batchFuture = batchExecutorService.submit(() -> {
```

Comment:
`BatchExecutorService` doesn't seem to be necessary.

Reply:
I asked too. https://github.com/apache/skywalking/pull/7153#discussion_r657171941